core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -231,10 +240,23 @@ public void run() {
 
   private ClientTransport obtainActiveTransport() {
     ClientTransport savedActiveTransport = activeTransport;
-    if (savedActiveTransport != null) {
+    // If we know there is an active transport and we are not in backoff mode, return quickly.
+    if (savedActiveTransport != null && !(savedActiveTransport instanceof InactiveTransport)) {
       return savedActiveTransport;
     }
     synchronized (lock) {
+      if (savedActiveTransport instanceof InactiveTransport) {
```

Comment:
`activeTransport` may have changed while entering the lock. Move this immediately after the `savedActiveTransport = activeTransport;` that is a couple lines below here.